### DISCUSSION TITLE
Composition and XRD Versioning

### DISCUSSION THREAD

**Discussion by blut** [crossplane/crossplane#2605] [Q&A] [answered]
Hello everyone

I'm currently working with a v1alpha1 version of a composition and have some breaking changes which I want to release as v1alpha2.
I've adjusted the XRD to include a v1alpha2, which I've set as referenceable: true and deployed a second composition, which is set as the new default composition in the XRD.
Now I'd expect any claim deployed with v1alpha1 to result in a composite v1alpha1 and any claim with v1alpha2 to result in a composite v1alpha2, however any v1alpha1 claim created is also accessible as a v1alpha2 claim, which then results in an invalid composite v1alpha2.

Workaround would be to use apiVersion as composition label and to add a compositionSelector to each claim.
Is there a better / intended way?

The following steps can be used to reproduce the behaviour:
To create the current state:
kubectl apply -f xrd-v1.yaml
kubectl apply -f composition-v1alpha1.yaml
kubectl apply -f existing-claim.yaml

Testing if composites can be processed:
kubectl describe compositeversioningtest | grep -e "^\(Events\|Name\):$" -A 8
-> Successfully composed resources

Upgrade with v1alpha2:
kubectl apply -f xrd.yaml
kubectl apply -f composition-v1alpha2.yaml

Testing:
kubectl apply -f test.yaml

Check if all expected compositions exist and can be processed:
kubectl describe compositeversioningtest | grep -e "^\(Events\|Name\):$" -A 8
-> Only claims created with v1alpha2 reach Successfully composed resources
-> v1alpha1 created resources (upgrading-test & exists-test) become incompatible with the selected resource

Additional comment:
Claims created with v1alpha1 cannot be deleted anymore, since crossplane does not remove the finalizer while the claims do not fulfill the new api of v1alpha2.
To delete manually edit the resource to contain the newly required parameter spec.parameters.name.

Attached files:
[xrd-v1.yaml.txt](https://github.com/crossplane/crossplane/files/7249897/xrd-v1.yaml.txt)
[composition-v1alpha1.yaml.txt](https://github.com/crossplane/crossplane/files/7249901/composition-v1alpha1.yaml.txt)
[existing-claim.yaml.txt](https://github.com/crossplane/crossplane/files/7249902/existing-claim.yaml.txt)
[xrd.yaml.txt](https://github.com/crossplane/crossplane/files/7249985/xrd.yaml.txt)
[composition-v1alpha2.yaml.txt](https://github.com/crossplane/crossplane/files/7249913/composition-v1alpha2.yaml.txt)
[test.yaml.txt](https://github.com/crossplane/crossplane/files/7249919/test.yaml.txt)

**Comment by jbw976** [comment]
Thanks @blut for the detailed repro steps and expectations for this scenario.  We'll take a look into this and get back with some guidance on this soon! :)

**Comment by turkenh** [comment, marked as answer]
**Independent of XRDs and Crossplane concepts:**

Kubernetes allows serving multiple versions of custom resources (e.g. v1alpha1, v1alpha2) but only marks one of them as storage version (e.g. v1alpha2). 

- https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#specify-multiple-versions
- https://book.kubebuilder.io/multiversion-tutorial/api-changes.html#storage-versions

So, when you added a new version to your XRD, CRD for your claim, `versioningtests.versioning.test` correctly updated with the two versions (`v1alpha1` and `v1alpha2`) but only `v1alpha2` is marked with `storage: true`. At this point all new resources will be stored with `v1alpha2` schema, even if you specify as `apiVersion: versioning.test/v1alpha1` in your yaml. For the existing ones, you can get with old schema until you made your first (whatever) update to your resource. After the first update, it will be converted and stored with the new schema (i.e. `v1alpha2`) and you will lose fields of the old schema if they do not exist in the new one.

Kubernetes allows you to handle such cases by specifying a conversion strategy in your CRD spec, which could either be `None` or `Webhook` as of today. See [webhook conversion](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#webhook-conversion) for further details.

**Coming back to Crossplane world:**

XRD's [supports multiple versions](https://github.com/crossplane/crossplane/pull/1884) but do not support specifying a conversion strategy as of today (just created [an issue](https://github.com/crossplane/crossplane/issues/2608) for that):

> Note that all versions must (by convention) have identical schemas; generated CRDs always use the 'None' conversion strategy to convert between the schemas of different CR versions. Crossplane may support other conversion strategies (e.g. webhooks or declarative conversion) in future.

So, I believe you can solve your problem by making a backward-compatible change for your new version (e.g. do not remove or rename an existing field, rather deprecate). This would prevent information for the old version from getting lost when stored with the new version.

**Reply by blut** [reply]
Thanks for your insight.
We will work with compositionSelectors for now, until we can include convsersion strategies.
